src/main/java/net/fabricmc/loom/providers/MappingsProvider.java

```diff
@@ -66,50 +112,133 @@ public void provide(DependencyInfo dependency, Project project, LoomGradleExtens
+		reorderMappings(unmergedIntermediary, invertedIntermediary, "intermediary", "official");
+		Path unorderedMergedMappings = Paths.get(mappingsStepsDir.toString(), "unordered-merged.tiny");
+		project.getLogger().lifecycle(":merging");
+		mergeMappings(invertedIntermediary, unmergedYarn, unorderedMergedMappings);
+		reorderMappings(unorderedMergedMappings, tinyMappings.toPath(), "official", "intermediary", "named");
+	}
+
+	private void reorderMappings(Path oldMappings, Path newMappings, String... newOrder) {
+		Command command = new CommandReorderTinyV2();
+		String[] args = new String[2 + newOrder.length];
+		args[0] = oldMappings.toAbsolutePath().toString();
+		args[1] = newMappings.toAbsolutePath().toString();
+		System.arraycopy(newOrder, 0, args, 2, newOrder.length);
+		runCommand(command, args);
+	}
+
+	private void mergeMappings(Path intermediaryMappings, Path yarnMappings, Path newMergedMappings) {
+		try {
+			Command command = new CommandMergeTinyV2();
+			runCommand(command, intermediaryMappings.toAbsolutePath().toString(),
+							yarnMappings.toAbsolutePath().toString(),
+							newMergedMappings.toAbsolutePath().toString(),
+							"intermediary", "official");
+		} catch (Exception e) {
+			throw new RuntimeException("Could not merge mappings from " + intermediaryMappings.toString()
+							+ " with mappings from " + yarnMappings, e);
+		}
+	}
+
+	private void suggestFieldNames(MinecraftProvider minecraftProvider, Path oldMappings, Path newMappings) {
+		Command command = new CommandProposeFieldNames();
+		runCommand(command, minecraftProvider.MINECRAFT_MERGED_JAR.getAbsolutePath(),
+						oldMappings.toAbsolutePath().toString(),
+						newMappings.toAbsolutePath().toString());
+	}
+
+	private void runCommand(Command command, String... args) {
+		try {
+			command.run(args);
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
 	}
 
-	public void initFiles(Project project) {
+	private void initFiles(Project project) {
 		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
-		MAPPINGS_DIR = new File(extension.getUserCache(), "mappings");
+		mappingsDir = extension.getUserCache().toPath().resolve("mappings");
+		mappingsStepsDir = mappingsDir.resolve("steps");
 
-		MAPPINGS_TINY_BASE = new File(MAPPINGS_DIR, mappingsName + "-tiny-" + minecraftVersion + "-" + mappingsVersion + "-base");
-		MAPPINGS_TINY = new File(MAPPINGS_DIR, mappingsName + "-tiny-" + minecraftVersion + "-" + mappingsVersion);
-		MAPPINGS_MIXIN_EXPORT = new File(extension.getProjectBuildCache(), "mixin-map-" + minecraftVersion + "-" + mappingsVersion + ".tiny");
+		tinyMappingsWithoutEnums = mappingsDir.resolve(mappingsName + "-tiny-" + minecraftVersion + "-" + mappingsVersion + "-base");
+		tinyMappings = mappingsDir.resolve(String.format("%s-tiny-%s-%s.tiny", mappingsName, minecraftVersion, mappingsVersion)).toFile();
```

Comment:
Should this have -v2 if v2?

Reply:
I need to check, previously there was no point because `mappingsName` was different

Reply:
yeah `mappingsName` got mutated to match, anyway not relevant anymore